src/plonk/circuit.rs

```diff
@@ -1072,11 +1156,280 @@ impl<F: Field> ConstraintSystem<F> {
+    /// polynomials for those columns. Finally, an internal map is updated to
+    /// find which fixed column corresponds with a given `Selector`.
+    ///
+    /// Do not call this twice. Yes, this should be a builder pattern instead.
+    pub(crate) fn compress_selectors(mut self, selectors: Vec<Vec<bool>>) -> (Self, Vec<Vec<F>>) {
+        if selectors.is_empty() {
+            return (self, vec![]);
+        }
+        // The length of all provided selectors must be the same.
+        let n = selectors[0].len();
+        assert!(selectors.iter().all(|a| a.len() == n));
+        // and the number of provided selectors must be the same.
+        assert_eq!(selectors.len(), self.num_selectors);
+
+        // Compute the maximal degree of every simple selector. We only consider
+        // the expressions in gates, as lookup arguments cannot support simple
+        // selectors.
+        let mut degrees = vec![0; selectors.len()];
+        for expr in self.gates.iter().flat_map(|gate| gate.polys.iter()) {
+            let degree = expr.degree();
+            if let Some(selector) = expr.extract_simple_selector() {
+                degrees[selector.0] = max(degrees[selector.0], degree);
+            }
+        }
+
+        // Any elements of `degrees` which still has zero must either be a
+        // simple selector or appear in no columns, so we don't need to compute
+        // anything for those. They will appear on their own.
+        let complex_selectors = degrees.iter().map(|&i| i == 0).collect::<Vec<_>>();
+
+        // Compute the exclusion matrix that has (j, k) = true if selector j and
+        // selector k conflict -- that is, they are both enabled on the same
+        // row. This matrix is symmetric so we only need to store the lower
+        // diagonal.
+        let mut exclusion_matrix = (0..self.num_selectors)
+            .map(|i| vec![false; i])
+            .collect::<Vec<_>>();
+
+        // Iterate over each of the selector columns
+        for (i, rows) in selectors.iter().enumerate() {
+            // Ignore exclusions with complex selectors
+            if complex_selectors[i] {
+                continue;
+            }
+
+            // Iterate over the rows, filtering for rows that are set
+            for set_index in rows.iter().enumerate().filter(|a| *a.1).map(|a| a.0) {
+                // Loop, looking only at the selectors previous to this one
+                for (j, other_selector) in selectors.iter().enumerate().take(i) {
+                    // Ignore exclusions with complex selectors
+                    if complex_selectors[j] {
+                        continue;
+                    }
+
+                    // Look at what selectors are active at the same row
+                    if other_selector[set_index] {
+                        // Mark them as incompatible
+                        exclusion_matrix[i][j] = true;
+                    }
+                }
+            }
+        }
```

Comment:
Let n be the number of rows.
Let s be the number of selector columns (assume for now there are no complex selectors).
Let q[r] be the number of enabled selectors on row r.

This algorithm costs O(n.s + sum_r {q[r].s}) time, whereas the one in the Python prototype takes O(n.s + sum_r {q[r]^2}) time, where q[r] will typically be smaller than s. Here's the Python code adapted for column-major order: https://gist.github.com/daira/7c9a2a372a89606ce1d2207e7e46ba79